### PR TITLE
Modernize module, prep for Terraform 0.12 and TPG 2.0.0

### LINKED ISSUE #10: Input variables listed in the README.me aren't used
network and subnetwork are listed as input variables for terrform-google-vault but neither of them is used. when attempting to use them they are ignored
```module "vault" {
  source = "github.com/GoogleCloudPlatform/terrform-google-vault"
  project_id           = "${var.project_id}"
  region               = "${var.region}"
  zone                 = "${var.zone}"
  storage_bucket       = "${var.storage_bucket}"
  kms_keyring_name     = "${var.kms_keyring_name}"
  network              = "foo"
  subnetwork           = "bar"
}

### LINKED ISSUE #15: Feature Request: Auto scaling group
I was wondering what extra config would be required to add a load balancer and an ASG that would scale out automatically based on CPU usage ?

### LINKED ISSUE #19: seal-status: x509: certificate signed by unknown authority
Following solution guide https://cloud.google.com/solutions/using-vault-for-secret-management, but I got the following error when running the vault unseal command.

Error checking seal status: Get https://127.0.0.1:8200/v1/sys/seal-status: x509: certificate signed by unknown authority

I made sure these variables were set, but I must be missing something.

export VAULT_ADDR=https://127.0.0.1:8200
export VAULT_CACERT=/etc/vault/vault-server.ca.crt.pem
export VAULT_CLIENT_CERT=/etc/vault/vault-server.crt.pem
export VAULT_CLIENT_KEY=/etc/vault/vault-server.key.pem

### LINKED ISSUE #23: Accessing the vault server   from another service
I followed the example to setup the vault server as is . Now I want it to be exposed to a particular service to store and retrieve secrets. How do I go about it. What configurations should I change. 

Thank you in advance

### LINKED ISSUE #30: terraform init failing with Invalid default value for variable "tls_ca_subject"
terraform plan
Error: Invalid default value for variable
  on ../../variables.tf line 76, in variable "tls_ca_subject":
  76:   default = {
  77:     common_name         = "Example Inc. Root"
  78:     organization        = "Example, Inc"
  79:     organizational_unit = "Department of Certificate Authority"
  80:     street_address      = ["123 Example Street"]
  81:     locality            = "The Intranet"
  82:     province            = "CA"
  83:     country             = "US"
  84:     postal_code         = "95559-1227"
  85:   }
This default value is not compatible with the variable's type constraint: all
map elements must have the same type.

changing street_address type to String fixing the issue.

### LINKED ISSUE #34: deprecations
Warning: module.vault.module.vault-server.google_compute_instance_group_manager.default: "auto_healing_policies": [DEPRECATED] This field is in beta and will be removed from this provider. Use it in the the google-beta provider instead. See https://terraform.io/docs/providers/google/provider_versions.html for more details.

Warning: module.vault.module.vault-server.google_compute_instance_group_manager.default: "rolling_update_policy": [DEPRECATED] This field is in beta and will be removed from this provider. Use it in the the google-beta provider instead. See https://terraform.io/docs/providers/google/provider_versions.html for more details.

Warning: module.vault.module.vault-server.google_compute_instance_template.default: "network_interface.0.address": [DEPRECATED] Please use network_ip

### LINKED ISSUE #35: Error 403: The caller does not have permission, forbidden`
`error is googleapi: Error 403: The caller does not have permission, forbidden` when calling iam api, while I am owner of the project.

https://github.com/hashicorp/terraform/issues/13109#issuecomment-297719601

### LINKED ISSUE #36: resource "google_project_iam_policy" causes deleting iam policy
```
{resource "google_project_iam_policy" "vault" { project = "${var.project_id}" policy_data = "${data.google_iam_policy.vault.policy_data}" }

```
 
Which makes problem!that when you destroy vault using terraform it will restore the project iam policy to the state that was before destroying.

That means all the new iam policies that was created during the period that Vault was created by terraform and till the the time vault was destroyed are gonna be gone !

### LINKED ISSUE #38: Fields for google provider [REMOVED]
Note this issue is related to #34 but I am getting a slightly different error. 

**Problem**: When attempting to use the [vault-on-gce example](https://github.com/GoogleCloudPlatform/terraform-google-vault/tree/master/examples/vault-on-gce) I get the following errors after executing `terraform plan`. 

```
Error: module.vault.module.vault-server.google_compute_instance_group_manager.default: "auto_healing_policies.0.initial_delay_sec": [REMOVED] This field is in beta. Use it in the the google-beta provider instead. See https://terraform.io/docs/providers/google/provider_versions.html for more details.

Error: module.vault.module.vault-server.google_compute_instance_group_manager.default: "rolling_update_policy": [REMOVED] This field is in beta. Use it in the the google-beta provider instead. See https://terraform.io/docs/providers/google/provider_versions.html for more details.

Error: module.vault.module.vault-server.google_compute_instance_template.default: "network_interface.0.address": [REMOVED] Please use network_ip
```

**Attempted fixes**: 

1)  I attempted to execute the terraform execution in the context of google cloud shell to see if there was perhaps something unique about my local environment but that did not change the outcome. 

2) Reading the error and associated documentation link it appears to me that I should be able to switch from the `google` provider to the `google-beta` provider to resolve this issue. I edited the [following](https://github.com/GoogleCloudPlatform/terraform-google-vault/blob/b11e8f44bbb28c6936a703fba6863ad5b681c772/examples/vault-on-gce/main.tf#L29) line to be `google-beta`.

This did not resolve the issue. I still see the `google` provider being provisioned during `terraform init`. I am fairly new to terraform and in reading about the plugin system it seems there isn't a straightforward way to remove old providers. 

To help mitigate the potential of an old `google` version of the provider being the issue I started over again in a new folder, as terraform will provision providers for only the [current working directory](https://www.terraform.io/docs/configuration/providers.html#initialization).  However when I execute `terraform init` I still see the google provider being initialized.

```
* provider.google: version = "~> 2.0"
* provider.google-beta: version = "~> 2.0"
```
Attempting `terraform plan` after the initialization in a new folder does still trigger the same error.

I searched through the remainder of the codebase in the hope that I would find a call to the original google provider but I haven't been able to find anything. This is literally my first time using terraform, I very well may be chasing the wrong issue. 

**System information**: 

* osx: 10.14.3
* Terraform: v0.11.11

### PR DESCRIPTION
**Pull request by sethvargo**
This PR is large, but there isn't a great way to break it down.

- Fixes #40, fixes #38, fixes #34, fixes #30 - tested with TF 0.12 and TPG 2.0.0
- Fixes #39 - stackdriver monitoring and logging are setup
- Fixes #35, fixes #36 - iam permissions are always enabled, switched to non-authoritative so users can add their own iam perms
- Fixes #32 - obsolete
- Fixes #29 - service account name is a variable; use terraform workspaces for multiple environments
- Fixes #25 - expose service account email in output, keys are no longer generated (obsolete) 
- Fixes #23 - documented in FAQ (TL;DR access from the LB always)
- Fixes #19 - TLS LB is valid for the cert
- Fixes #15 - regional autoscaler in place
- Fixes #12 - obsolete as vault runs in a dedicated network now
- Fixes #9 - instances can assign custom tags, labels, metadata
- Fixes #10, fixes #1 - inputs, outputs, and resources are auto-parsed by the Terraform module registry

### CODE REVIEW

**Comment by LandonSchropp** [comment]
Thanks again Seth! You're the best. 😎

**Comment by sethvargo** [comment]
We need to publish this to the module registry then too

File: scripts/gcpkms-encrypt.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+# Extract args into shell variables
+eval "$(jq -r '@sh "ROOT=\(.root) DATA=\(.data) PROJECT=\(.project) LOCATION=\(.location) KEYRING=\(.keyring) KEY=\(.key)"')"
+
+# If no root or data was given, error
+[[ "${ROOT}" == "null" ]] && echo "Missing ROOT!" && exit 1
+[[ "${DATA}" == "null" ]] && echo "Missing DATA!" && exit 1
+
+# Calculate the md5 of the data - used to lookup
+SIG=$(echo -n "${DATA}|${PROJECT}|${LOCATION}|${KEYRING}|${KEY}" | md5 | cut -d ' ' -f1)
```
Comment (danisla) [inline scripts/gcpkms-encrypt.sh]: This did not work in cloud shell.

Here is a proposed cross-platform fix:

```bash
# Calculate the md5 of the data - used to lookup
MD5SUM=$(command -v md5 || true)
[[ -z "${MD5SUM}" ]] && MD5SUM=$(command -v md5sum || true)
[[ -z "${MD5SUM}" ]] && echo "Missing command: 'md5' or 'md5sum'" >&2 && exit 1

SIG=$(echo -n "${DATA}|${PROJECT}|${LOCATION}|${KEYRING}|${KEY}" | ${MD5SUM} | cut -d ' ' -f1)
```

File: examples/vault-on-gce/README.md [hunk 2]
```diff
@@ -1,197 +1,189 @@
...
+# HashiCorp Vault on Compute Engine
 
-**Figure 1.** *diagram of Google Cloud resources*
+This example shows how to run a highly-available HashiCorp Vault cluster on
+Google Compute Engine.
 
-![architecture diagram](./diagram.png)
 
-## Create the Cloud KMS KeyRing for asset encryption:
+## Setup environment
```
Comment (danisla) [inline examples/vault-on-gce/README.md]: Add a step to install terraform as well?

File: README.md [hunk 3]
```diff
@@ -1,53 +1,214 @@
...
+1. Execute Terraform:
+
+    ```
+    $ terraform apply
+    ```
+
+1. Configure your local Vault binary to communicate with the Vault server:
+
+    ```
+    $ export VAULT_ADDR="$(terraform output vault_addr)"
```
Comment (danisla) [inline README.md]: Should be:

```bash
export VAULT_ADDR="$(terraform output -module vault vault_addr)"
```

File: scripts/gcpkms-encrypt.sh (hunk 1 above)
Comment (sethvargo) [reply scripts/gcpkms-encrypt.sh]: Turns out md5sum actually works everywhere and md5 is just an alias on darwin. Updated.

**Pull request merged**